### ISSUE TITLE
Context propagation is not working for WebSocket.

### ISSUE THREAD

**Issue by SetoKaiba** [quarkusio/quarkus#30994]
### Describe the bug

The context propagation is not working for WebSocket.
Please referred this post here.
https://github.com/quarkusio/quarkus/discussions/30905#discussioncomment-4898246
@sberyozkin suggest that create a simple reproduce project to validate this. But I have no idea what bean can be injected to validate. Or maybe anyone other reproduce this.
From the thread name we can find the difference.

### Expected behavior

The context propagation should work for WebSocket. The security identity should be the same. Should not lost the authentication.
Here's a snapshot that context propagation is working
![image](https://user-images.githubusercontent.com/159540/217496455-a188cb22-e638-4fd5-834f-b9e1aa91a77a.png)
![image](https://user-images.githubusercontent.com/159540/217496491-5b87a6d2-1f2c-4f6a-94e1-aa9a16dd39f8.png)

### Actual behavior

The context propagation is not working for WebSocket. The security identity become anonymous.
Here's a snapshot that context propagation is not working.
![image](https://user-images.githubusercontent.com/159540/217496550-9e2a3a52-43e2-4102-afba-74b09c5ed2d4.png)
![image](https://user-images.githubusercontent.com/159540/217496596-4a72c1a2-a87f-4018-a977-f0c098529843.png)

### How to Reproduce?

Just use the Actual behaviour code to reproduce. But it may need some settings with security.

I have no idea how to reproduce this without security.

### Output of `uname -a` or `ver`

Microsoft Windows [版本 10.0.19045.2486]

### Output of `java -version`

openjdk 11.0.12 2021-07-20 OpenJDK Runtime Environment Microsoft-25199 (build 11.0.12+7)          OpenJDK 64-Bit Server VM Microsoft-25199 (build 11.0.12+7, mixed mode)

### Quarkus version or git rev

3.0.0.Alpha3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

------------------------------------------------------------ Gradle 7.5.1 ------------------------------------------------------------  Build time:   2022-08-05 21:17:56 UTC Revision:     d1daa0cbf1a0103000b71484e1dbfe096e095918  Kotlin:       1.6.21 Groovy:       3.0.10 Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021 JVM:          11.0.15 (GraalVM Community 11.0.15+10-jvmci-22.1-b06) OS:           Windows 10 10.0 amd64

**Comment by SetoKaiba** [comment]
Any update?

**Comment by SetoKaiba** [comment]
@FroMage @manovotn @sberyozkin Anyone helps please?
I created an issue as @sberyozkin suggest. But no one here?

**Comment by geoand** [comment]
@manovotn any idea?

**Comment by manovotn** [comment]
IIUIC, this issue is purely about propagation in websockets, correct? You are using security but it's not linked to the actual problem.
I am not familiar with how or if websockets do any propagation under the hood (I don't think they do) but the idea to use `ManagedExecutor` to trigger it manually seems correct. If you are not seeing it propagate, I'd start by checking if there is any context active before you attempt the propagation.

I.e. in your `@OnOpen public void onOpen(..) {}`, you could check if there is CDI req. context active.
This can be done by:
> Arc.container().requestContext().isActive()

I'd also love to see how you obtain and use `ManagedExecutor`.

@SetoKaiba please link a project or attach a reproducer that we could debug. Doesn't matter if there is a bit of security in it as long as this is reproducible :)

**Comment by SetoKaiba** [comment]
I don't remember that how can I make it work for REST. Now, both are not working. Maybe I miss something. It's somehow long time ago. Here's a reproduce project. What I want is in the sub thread the identity security info is still available.
https://github.com/SetoKaiba/reproduce

**Comment by manovotn** [comment]
Thanks, I will take a look at it later this week and report back if I can spot anything.

**Comment by manovotn** [comment]
@SetoKaiba how do you run the reproducer? Performing a build and Quarkus dev mode boot (`./gradlew quarkusDev`), I am getting an error during startup.

Meanwhile, @sberyozkin is security context propagation in Quarkus bound to CDI context propagation or does it have its own mechanism (i.e. for instance in MP CP that would translate to separate `ThreadContextProvider`)?

**Comment by manovotn** [comment]
@SetoKaiba I can see the popup now and login. But `Request REST with Bearer` does nothing and the remaining three buttons still don't seem to work correctly because I am seeing anon identity and null web token with all of them:

> 2023-04-03 12:51:09,351 INFO  [net.kai.LiveServerEndpoint] (vert.x-eventloop-thread-3) io.quarkus.security.runtime.AnonymousIdentityProvider$1@f965002
2023-04-03 12:51:09,352 INFO  [net.kai.LiveServerEndpoint] (vert.x-eventloop-thread-3) io.quarkus.oidc.UserInfo@5369a98d
2023-04-03 12:51:09,353 INFO  [net.kai.LiveServerEndpoint] (vert.x-eventloop-thread-3) io.smallrye.jwt.auth.cdi.NullJsonWebToken@711641d3

**Comment by manovotn** [comment]
@SetoKaiba even without the example properly working, I think I can see why the CDI context isn't propagating for the `@OnOpen` method and I suppose the REST method will have the exact same problem.

The setup you are doing with `ManagedExecutor` is technically correct and you will have req. context active inside the task you submit (which you are correctly testing). However, it will be a different req. context -  you can see this if you ask `Arc` for the underlying context state and then log it like this:

>      InjectableContext.ContextState state = Arc.container().requestContext().getState();
>      logger.info(state + " is valid: " + state.isValid());

Do the same from inside the `ManagedExecutor` task you are submitting and you can see the difference.

The problem is that you are trying to use CDI context after it has already ended (and been invalidated) on the original thread.
That is not permitted because otherwise you could extend request context life indefinitely and CDI would not know when to terminate such context as threads could just endlessly pass it around. Therefore, during context propagation, Arc checks the `ContextState` object and if it is invalid, it will instead create a fresh req. context.

Here is a Gist with two variations of your code that you can try and see what I mean - https://gist.github.com/manovotn/2fe271caab8f0f9e74417eebb162197a

**Comment by manovotn** [comment]
> Do, do you mean that the context propagation will only work without the request ended? 

If you intend to use whatever was stored in the previous request context, then yes. Note that other contexts such as application and singleton carry over their information into other threads and can be accessed from there.
Generally speaking, the context propagation of CDI req. context is deliberately done in a way so that the original thread "owns" the req. context state and determines when it starts and ends. Any parallel threads that it gets propagated into have to check if the state is valid by the time they try to use it and act accordingly.

> The onOpen can't be blocked for a long time. 

Understood, I only added the blocking to the gist code to showcase how it behaves.

> Does the only way is to read all the data out and pass to a method doesn't need a request context?

In short yes, you most likely need to extract data from any req. context beans and process them separately since the req. context will end by the time you get to processing.
More precisely, if you intend on offloading the work to another task via executor, then such task shouldn't rely on using *the same req. context state*. It can still require active req. context in order to create some beans, but those will be new instances (which is exactly what happens in your current code).
